### PR TITLE
[#2776] Rate limiter for request calls to node server

### PR DESCRIPTION
**Pull request by LetItRock**
**Status**
- [x] WIP
- [ ] Ready for review
- [ ] Needs testing

**Deploy after merge (delete what needn't be deployed)**
- api
- hyperion (frontend)

**Release notes for users (delete if codebase-only change)**
- express middleware which will limit request calls to API for authenticated/anonymous user within a time frame; it will create a key in redis "rateLimiter:limit:id"; where id for authenticated user is his "id" and for anonymous user "ip address";

We should also consider values for max requests and for time frame.
Do we want to implement same middleware for other servers? Which one?

### CODE REVIEW

**Review by mxstbr — COMMENTED** [review]
This looks amazing! 😍 We definitely also want to use this in hyperion, although hyperion probably needs stricter limits than the API. (since it's rendering requests take a while longer than the API responses)

I'm not very familiar with rate limiting, so I'm not 100% sure what the limiting values should be. Any ideas?

**Comment by LetItRock** [comment]
Ok, I'll also implement this in hiperion. Me neither, maybe we should look into statistics/analytics?

**Comment by mxstbr** [comment]
Yeah, maybe have a look around/google what other people are doing, I'm really not sure.

**Comment by LetItRock** [comment]
I've implemented this also for hyperion ;)

**Comment by brianlovin** [comment]
@mxstbr are you cool owning the testing on this one and deploying to alpha?

**Review by mxstbr — COMMENTED** [review]
I looked around, and [Trello allows 300 requests per 10 seconds to their API](https://help.trello.com/article/838-api-rate-limits). (which is an avg of 30 requests per second)

That seems very high to me—our frontend when used by a legitimate user does at most probably like 40-50 requests in a 10 second window. (and that's when clicking around like a mad man and not doing anything actually useful)

Let's try restricting the API to 100 req per 10 sec? (avg of 10 req per second, still a ton)

----

Hyperion will need a way lower limit because it's a lot more expensive to render a page. Thinking out loud about this, no normal user should ever have to make more than 5 req/sec, so let's try restricting hyperion to 50 requests in 10 sec?

**Comment by chrbala** [comment]
Make sure to look into the VPN case if you're grouping by IP. It's possible that there might be a lot of users on the same IP if they're all on the same VPN. 

We occasionally get rate limited on sites because of the number of users on our corporate network.

**Review by mxstbr — COMMENTED** [review]
We should make sure that hyperion still serves cached responses for logged-out users, and only rate limit new renders.

**Comment by LetItRock** [comment]
@brianlovin I'm not added to "now" team. I can try to own testing and deploying to alpha if I will have permissions to do that.

@mxstbr I think the same way as you that 100 its a lot, so lets restrict API calls to 100 per 10 sec for now. 
I was testing hyperion and figured out that 1 page refresh for ex. when I'm on home page makes 7 calls to RateLimiter... is that related to resource downloading?

@chrbala good catch, could you point me how can we archive this?

**Comment by mxstbr** [comment]
> I was testing hyperion and figured out that 1 page refresh for ex. when I'm on home page makes 7 calls to RateLimiter... is that related to resource downloading?

Arg I totally forgot about the static files, damn. Can we add the rate limiting middleware after the `express.static` one?

**Comment by LetItRock** [comment]
@mxstbr sure... and I will check that later today ;)

File: api/routes/middlewares/index.js [hunk 1]
```diff
@@ -50,6 +50,10 @@ import passport from 'passport';
 middlewares.use(passport.initialize());
 middlewares.use(passport.session());
 
+// rate limiter middleware to reduce API calls to server within time frame
+import rateLimiter from 'shared/middlewares/rateLimiter';
+middlewares.use(rateLimiter('api', 2500, 3600000));
```
Comment (andreasklinger) [inline api/routes/middlewares/index.js]: outsider pov: 
imo bad (aka magic) numbers b/c i need to look at source what they mean 

`rateLimiter('api', MAX_REQUESTS, ONE_HOUR)`

normally i'd say only name constants and variables by WHAT not HOW but i think stuff like ONE_HOUR is an exception b/c you would do it with syntax like `1.hour` if you could - so it's more a workaround to make it readable in JS

Comment (mxstbr) [reply api/routes/middlewares/index.js]: Yeah totally, I don't think @LetItRock had thought those numbers through yet

Comment (LetItRock) [reply api/routes/middlewares/index.js]: @andreasklinger yeah for sure I will extract those values to more readable constants... till today I didn't know what exact values they gonna be... and we are still debating on values for hyperion...

**Comment by LetItRock** [comment]
@mxstbr I've moved rate limiter middleware code below "cache"... and we now get few requests less to rate limiter... those are:
- / 
- /static/js/bootstrap.js
- /undefined/img/favicon.ico

I think that it should be possible to reduce those two last right? :P 
......................

Another thing is VPN/Proxy, as far as I know they basically almost the same thing (VPN encrypts data - proxy not)... I know that proxies are using "X-Forwarded-For" header to kind of "store" real user IP address... but I'm not sure about VPN do they handle requests same way with same header? Can someone confirm that?
Our RateLimiter middleware is using request-ip library which handles "X-Forwarded-For" and other specific headers from [docs](https://github.com/pbojinov/request-ip)

**Comment by mxstbr** [comment]
> - /static/js/bootstrap.js
> - /undefined/img/favicon.ico

What on earth is that second one?! The first one should be served with `build/static/js/bootstrap.js` it's interested that the static middleware doesn't see that one...

**Comment by mxstbr** [comment]
Oh actually, that first one should be fixed in `src/views/navbar/index.js` and `src/views/pages/components/nav.js`—it says `href={${process.env.PUBLIC_URL}/img/favicon.ico}` in both those places, but PUBLIC_URL is no longer defined so let's try and change that to `/img/favicon.ico`.

**Comment by LetItRock** [comment]
@mxstbr 
So the issue with favicon is fixed... for the bootstrap requests I see strange thing is happening... it basically returns index.html page for /static/js/bootstrap.js request 😄 thats why it goes through RateLimiter... 
I've checked package.json and I don't see any dependency to bootstrap... but in config-overrides we have CommonChunksPlugin which creates the bootstrap.js chunk for us... but I've checked that file from served file and it only has webpack config inside... no bootstrap code...

---
@mxstbr @brianlovin And I need someone to answer my question regarding to VPN like few messages above...

**Comment by mxstbr** [comment]
- I'm not sure about the VPN thing. Would you mind googling around for the best way to solve that?
- With `bootstrap.js`, if you run `yarn run build:web` and then check the `build/static/js` folder you'll see that it contains the `bootstrap.js` file. It's weird that it would not be served though?

**Comment by LetItRock** [comment]
@mxstbr ok I will try to find something...

---
@mxstbr I've also checked that bootstrap file and it only has webpack specific code, it doesn't have anything related to bootstrap... maybe its some leftover... not sure... because I've tried to remove that chunk config from config-overrides and it was working fine for local development config... 

but on hyperion side I saw something strange happening first application was rendered normally and then after few second it disappeared... have no idea why its happening...
Also I got after those changes new request in RateLimiter for "/static/js/Splash.chunk.js"... weird... Its somehow related to routes.js file...

**Comment by LetItRock** [comment]
@mxstbr finally I got it how it works on hyperion side: every route change for ex. "/login", "/about" renders page on server side and then caches it in redis if user is not authenticated... 
Splash.chunk.js thats basically Pages component which renders different components based on routes... 
RateLimiter in hyperion will count every route request and "*.chunk.js" files for authenticated users... it will associate "user id" with number of possible requests in time frame... 
but for those who are not authenticated it will count only first time (by IP) and it will render that page and save in cache... every other request comming from unauthenticated users will be taken from cache... thats in hyperion case...

VPN - it looks like its not possible to access "real user IP" because VPN its like separate private network on your computer and all the requests are going through VPN server which has one IP address for all users connected to that private network... so on our end we will se VPN server IP address... thats my understanding...

Github has implemented rate limiting for API in second way: 5000 requests for authenticated user per hour, 60 requests for unauthenticated users per hour (associated with IP) [link](https://developer.github.com/v3/#rate-limiting);

StackExchange: 30 requests per second for application (associated by IP), if number is higher then they give ban for 30 sec... they also have daily limit for application without access_token to 10000, this number is given for all applications from same IP address... if the application has access_token that means its associated with user and it will have its app/user limit 10000 requests per day [link](https://api.stackexchange.com/docs/throttle);

Twitter: only limits authenticated users/applications and they have different limits to different endpoints [link](https://developer.twitter.com/en/docs/basics/rate-limiting), [link](https://developer.twitter.com/en/docs/basics/rate-limits);

SoundCloud: they calculate rate limits based on clients application client id and its 15000 per day [link](https://developers.soundcloud.com/docs/api/rate-limits);

Facebook [link](https://developers.facebook.com/docs/graph-api/advanced/rate-limiting) and Youtube [link](https://developers.google.com/youtube/v3/getting-started#quota) they have more compicated rate limits

**Comment by mxstbr** [comment]
Thanks for that in-depth research @LetItRock! Thoughts:

- Given that cached requests don't count towards the rate limit, and that every unauthenticated request to hyperion will respond with a cached request after the first time, it seems fine to keep that limit rather low even with VPNs since they'll get cached responses anyway.
- API queries really aren't as expensive to resolve, in comparison to the rendering that hyperion does, so we can keep a lenient limit there so even unauthenticated users behind VPNs can browse the site.

With that in mind, what about the following limits:

- **API: 300 reqs/10 secs**—very lenient, no legitimate user even behind a VPN should ever crack that limit, but it still protects the API from too many requests
- **Hyperion: 50 reqs/20 secs**—there's no way any legitimate user ever opens 50 uncached pages and has hyperion render them in a 20 second window, not even behind a VPN, given that any page you open after the initial load will be served with the app shell locally.

**Comment by LetItRock** [comment]
Yeah those thoughts are more clear 😃 thanks 👍 
I think that those limits should be enough even for unauthenticated users through VPN...
we need to give it a try and then will adjust it on the need... do we have some analytics tool?

Maybe it would be good to hear opinion from someone else from team on this? 

I will change those limits after we will have final decision...

**Comment by mxstbr** [comment]
> we need to give it a try and then will adjust it on the need... do we have some analytics tool?

It'd be great to log to Sentry everytime somebody exceeds the rate limit with their IP or user ID and what they were requesting. Let's get those changes in and ship it!

**Comment by LetItRock** [comment]
Sorry for late update on that... I've adjusted values for rate limiter for API and Hyperion, also implemented logging to Sentry 🎉

File: hyperion/renderer/html-template.js [hunk 2]
```diff
@@ -72,7 +73,6 @@ export const getFooter = ({
       <script>window.__SERVER_STATE__=${serialize(state)}</script>
       <script>window.__DATA__=${serialize(data)}</script>
       <script defer="defer" type="text/javascript" src="https://cdn.polyfill.io/v2/polyfill.min.js"></script>
-      <script type="text/javascript" src="/static/js/bootstrap.js"></script>
```
Comment (mxstbr) [inline hyperion/renderer/html-template.js]: This needs to be there, at least in production!

File: index.html [hunk 3]
```diff
@@ -50,5 +51,5 @@
   </head>
   <body>
     <div id="root"></div>
-  <script type="text/javascript" src="/./static/js/bootstrap.js"></script><script type="text/javascript" src="/static/js/bundle.js"></script></body>
```
Comment (mxstbr) [inline index.html:53]: This needs to be there in production!

File: hyperion/renderer/html-template.js (hunk 2 above)
Comment (LetItRock) [reply hyperion/renderer/html-template.js]: ok, sorry I forgot to remove that...

File: index.html (hunk 3 above)
Comment (LetItRock) [reply index.html:53]: sry

**Review by mxstbr — DISMISSED** [review]
This looks great to me! :100: Thanks so much for digging into this.

/cc @brianlovin whaddaya think?

**Review by brianlovin — DISMISSED** [review]

**Comment by brianlovin** [comment]
This looks good to me overall; obviously we're shooting in the dark a bit regarding the limits and timeframes; @mxstbr would you be cool deploying this to alpha first and let's keep an eye on this for a while to make sure our limits feel good!

File: api/routes/middlewares/index.js [hunk 4]
```diff
@@ -50,6 +50,12 @@ import passport from 'passport';
 middlewares.use(passport.initialize());
 middlewares.use(passport.session());
 
+// rate limiter middleware to reduce API calls to server within time frame
+import rateLimiter from 'shared/middlewares/rateLimiter';
+const THREE_HUNDRED = 300;
```
Comment (alejandronanez) [inline api/routes/middlewares/index.js]: Hello @LetItRock,

What do you think about giving this a more _generic_ name, maybe something like `BOTTOM_LIMIT` and `TOP_LIMIT`? Imagine if you Spectrum decided to change `300` to `800`, then you'd have to change the variable name too.

Comment (brianlovin) [reply api/routes/middlewares/index.js]: That's a good point; could be like:
`REQUESTS_ALLOWED` and `TIMEFRAME` - generic enough, and maps to what the arguments are doing in the function?

Comment (LetItRock) [reply api/routes/middlewares/index.js]: sure I'll do that, you are both right its better to have here more generic name...

**Comment by mxstbr** [comment]
Good call @brianlovin—I'm deploying these changes to `alpha.spectrum.chat` now!

**Comment by mxstbr** [comment]
This was deployed to the API and will be out for hyperion shortly—let's keep an eye out on alpha and see if this breaks anything in the common users usecase.

**Comment by brianlovin** [comment]
Closed with #4647 <3

**Pull request closed without merging**